Update Poll.

jduong09 opened the pull request:
Functionality to update created poll.

- Currently name, description, location, and duration properly display in the inputs of the form.
- Availabilities needs to be displayed on frontend side for user to see polls.

jduong09 committed:
Added router for /polls. fetched poll info for user to update.

jduong09 committed:
Need to add state of start times.

jduong09 committed:
Changed how times were stored into the database to accomodate morning and evening data.

jduong09 committed:
Minor Changes to Update Subscription.

jduong09 commented:
### Changes
- Fixed time frame display when looking at the table.
- Fixed submit issue where form was not sending a patch request.
- Deleted javascript file that contained duplicate function.
### Current Issues

- [ ] When updating subscription, changing the duration input will update all previous meeting times.
- [ ] Submitting update does not close form.
- [ ] Clicking edit button does not change form submit button text.

jduong09 committed:
Added form reset when submitting form. Need to add form reset if closing form.

jduong09 committed:
Fixed issues documented in pull request.

jduong09 commented:
### Changes

- Submitting form properly closes form component and displays updated polls.
- Create/Update forms properly display correct text on buttons.
- Removing meeting times functions properly in database and frontend.

jduong09 committed:
Added ReadMe file.

ziggysauce commented on root/client/src/components/poll/pollForm.js:
```diff
@@ -29,9 +44,9 @@ export const PollForm = ({ handleSubmit }) => {
       <label htmlFor='inputAvailabilities'>Availabilities
         <input id='inputAvailabilities' name='pollAvailabilities' type='text' value={pollAvailabilities} readOnly="readOnly" hidden/>
-        <Table duration={pollDuration} setPollAvailabilities={setPollAvailabilities} pollAvailabilities={pollAvailabilities} />
+      <Table duration={pollDuration} setPollAvailabilities={setPollAvailabilities} pollAvailabilities={pollAvailabilities} />
```
hm kinda weird having a table inside of a label tag, eh?

ziggysauce commented on root/client/src/components/user/dashboard.js:
```diff
@@ -4,28 +4,51 @@ import { PollForm } from '../poll/pollForm';
-    await fetch('/polls', { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(formData) })
+    e.preventDefault();
+
+    if (formMethod === 'POST') {
+      await fetch('/polls', { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(formData) })
+    } else {
+      await fetch(`/polls/${editData.id}`, { method: 'PATCH', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(formData) })
+    }
```
realistically, if the `formMethod` is POST it should have an isolated POST logic. I believe we can make it so that the `formMethod` can be PATCH which would then trigger separate logic so you don't need to nest everything into an if/else here

jduong09 replied:
I removed the formMethod state. Used the editData date in order to pass the correct action and method to the handle submit function in dashboard, which then will make the correct fetch request.

jduong09 committed:
Finished logic for updating meeting times for creator of poll.

jduong09 committed:
Removed form method state, handled method/action through arguments.

jduong09 committed:
Moved table component outside of label tag.

jduong09 commented:
### Changes
- Trying to edit a poll that is created by the user should display the previous chosen availabilities correctly.
- Submitting changes to those availabilities should work.

jduong09 committed:
Removed unnecessary console.logs and comments.

ziggysauce commented on root/backend/server/api/routes/poll.js:
```diff
@@ -0,0 +1,80 @@
+          }
+        });
+      });
+      res.json({ response: polls });
+      res.end();
```
do you _need_ `res.end()` here? can you instead just send a status and the data like so?

```suggestion
      res.status(200).json({ response: polls });
```

ziggysauce commented on root/backend/server/api/routes/poll.js:
```diff
@@ -0,0 +1,80 @@
+      await poll.save();
+      res.end();
+    } catch(err) {
+      console.log(err);
```
should be sending an error here. something like:

```javascript
res.status(400).json({ ...some error alert or something here... });
```

ziggysauce commented on root/backend/server/api/routes/poll.js:
```diff
@@ -0,0 +1,80 @@
+    await Poll.findOneAndUpdate({ _id: req.params.pollId }, pollData);
+    res.end();
+  } catch(err) {
+    console.log(err);
```
same as above

ziggysauce commented on root/backend/server/api/routes/poll.js:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    await Poll.findOneAndUpdate({ _id: req.params.pollId }, pollData);
+    res.end();
```
should probably be sending a success response status here. something like:

```javascript
res.status(200).json({ ...some data here if needed... });
```

or if you don't need to send back any data:

```javascript
res.sendStatus(200);
```

ziggysauce commented on root/backend/server/api/routes/poll.js:
```diff
@@ -0,0 +1,80 @@
+      };
+    });
+
+  res.json({ pollData });
+  res.end();
```
same comment as above

ziggysauce commented on root/client/src/components/user/dashboard.js:
```diff
@@ -4,28 +4,46 @@ import { PollForm } from '../poll/pollForm';
+  const handleSubmit = async (e, formData, action, method) => {
+    e.preventDefault();
+    
+    await fetch(action, { method, headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(formData) })
```
you should wrap this in a try/catch to handle any unforeseen errors

ziggysauce commented on root/client/src/components/user/dashboard.js:
```diff
@@ -4,28 +4,46 @@ import { PollForm } from '../poll/pollForm';
+
+  const handleEdit = async (poll_uuid, e) => {
+    e.preventDefault();
+    await fetch(`/polls/${poll_uuid}/pollInfo`)
```
same comment as above

ziggysauce reviewed (changes requested):
just some syntax and minor changes. haven't tested this yet but will do so next

jduong09 committed:
Update root/backend/server/api/routes/poll.js

Co-authored-by: Dan Nguyen <dtn3423@gmail.com>

jduong09 commented:
### Features Necessary
- [ ] Need to remove old date/times from database when that date/time has passed.

jduong09 committed:
Made changes based on recent comments on res.end. Implemented handling or backend errors.

jduong09 commented:
### Changes
- Added helper function that will make frontend api requests based on action and method given.
- Implemented handling of backend errors, displaying them as console.log/console.error on frontend side.

ziggysauce reviewed (changes requested):
creating a poll works fine. editing however, has some weird behavior:

https://github.com/jduong09/doodle/assets/25065329/38094dc9-6db4-4b38-9bd1-e1fd8408656d

---

when messing with the time tables in general too, if i click certain time slots that override others, previous ones are stored somewhere even if they're not visible in the UI. check this out where i add a bunch, but as i remove them, the previous ones become visible

https://github.com/jduong09/doodle/assets/25065329/f2b41f69-1d92-42dd-b0c4-7d5d5536c314

---

there's a bunch of UI optimizations for creating/editing a poll but we can save that for a different PR as that's out of scope

jduong09 committed:
Moved the logic for checking to see if a timeblock was selected previously.

jduong09 committed:
CSS work on extra space in table.

jduong09 committed:
Added mouseenter and mouseleave events to display possilbe meeting time.

jduong09 commented:
## Update
- I believe I fixed the issue with selected poll times disappearing on the UI when selecting a poll time near the previous time.
- Added UI element for when highlighting a time, to better show which time the user will select.

jduong09 committed:
Fixed excessive spacing.

ziggysauce commented on root/client/src/components/poll/pollForm.js:
```diff
@@ -1,15 +1,32 @@
+    resetFormData();
+  }
+
+  console.log(pollAvailabilities);
```
remove console log when done testing

ziggysauce commented on root/client/src/components/user/dashboard.js:
```diff
@@ -1,31 +1,60 @@
+    e.preventDefault();
+    try {
+      const { response } = await apiRequest(action, { method, headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(formData) })
+      console.log(response);
```
remove when done testing

jduong09 replied:
This will stay here until i add notifications.

ziggysauce commented on root/client/src/components/user/dashboard.js:
```diff
@@ -1,31 +1,60 @@
+      setShowPollForm(false);
+      setEditData(null);
+    } catch(err) {
+      console.log(err);
```
remove when done testing. perhaps some client-side error alert would serve well here

ziggysauce commented on root/client/src/components/user/dashboard.js:
```diff
@@ -1,31 +1,60 @@
+      setEditData(pollData);
+      setShowPollForm(true);
+    } catch(err) {
+      console.log(err);
```
same comment ad above

ziggysauce commented on root/client/src/components/user/dashboard.js:
```diff
@@ -1,31 +1,60 @@
+        setPolls(polls);
+      })
+      .catch(err => {
+        console.error(err);
```
some comment as above

ziggysauce commented on root/client/src/util/tablehelpers.js:
```diff
@@ -26,18 +36,34 @@ export const classTopPosition = (minutes) => {
-    return 'h-75';
-  } else if (duration === '60') {
-    return 'h-100';
+  if (duration === 15) {
+    return 'h-15';
+  } else if (duration === 30) {
+    return 'h-30';
+  } else if (duration === 45) {
+    return 'h-45';
+  } else if (duration === 60) {
+    return 'h-60';
+  } else if (duration === 120) {
+    return 'h-120'
   }
```
hm since these are all equal in structure, can probs do something like

```suggestion
  return `h-${duration}`;
```

ziggysauce reviewed (commented):
minor UI thing, but you should let the highlighted sections take up the full width that way they don't appear as squished like they do here.

<img width="331" alt="Screen Shot 2023-11-15 at 5 53 42 PM" src="https://github.com/jduong09/doodle/assets/25065329/d9052f73-1b0c-4e88-b654-38db722efed0">

In general the table itself can take up more width to be accommodating. though i think on mobile you're going to run into some UI issues regardless and need to make adjustments. up to you if you want to tackle that in this branch or not since it's out of scope.

Otherwise, functionality this seemed to work just fine now.

jduong09 committed:
Made changes based on PR review.

jduong09 commented:
## Update
- Made changes based on recent PR comments made by Dan.
- Approved PR merge.
